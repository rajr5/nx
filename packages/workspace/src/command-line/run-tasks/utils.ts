import * as yargsParser from 'yargs-parser';
import * as yargs from 'yargs';
import { NxJson, readNxJson, readWorkspaceJson } from '../shared';
import { WorkspaceResults } from '../workspace-results';
import { ProjectGraphNode } from '../project-graph';

export interface Arguments<T extends yargs.Arguments> {
  nxArgs: T;
  targetArgs: T;
  overrides: any;
}

const ignoreArgs = ['$0', '_'];
export function splitArgs<T extends yargs.Arguments>(
  args: T,
  nxSpecific: (keyof T)[]
): Arguments<T> {
  const nx: any = {};
  const targetArgs: any = {};

  const overrides = yargsParser(args._.slice(1));
  delete overrides._;

  Object.entries(args).forEach(([key, value]) => {
    if (nxSpecific.includes(key as any)) {
      nx[key] = value;
    } else if (!ignoreArgs.includes(key)) {
      targetArgs[key] = value;
    }
  });

  return { nxArgs: nx, targetArgs, overrides };
}
export interface TaskArgs {
  projects: string[];
  target: string;
  all?: boolean;
}

export function projectHasTargetAndConfiguration(
  project: ProjectGraphNode,
  target: string,
  configuration?: string
) {
  if (
    !project.data ||
    !project.data.architect ||
    !project.data.architect[target]
  ) {
    return false;
  }

  if (!configuration) {
    return !!project.data.architect[target];
  } else {
    return (
      project.data.architect[target].configurations &&
      project.data.architect[target].configurations[configuration]
    );
  }
}

export interface Environment {
  nxJson: NxJson;
  workspaceJson: any;
  workspace: any;
}

export function readEnvironment(target: string): Environment {
  const nxJson = readNxJson();
  const workspaceJson = readWorkspaceJson();
  const workspace = new WorkspaceResults(target);

  return { nxJson, workspaceJson, workspace };
}
