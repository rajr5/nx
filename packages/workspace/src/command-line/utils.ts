import * as yargsParser from 'yargs-parser';
import * as yargs from 'yargs';

/**
 * These options are only for getting an array with properties of AffectedOptions.
 *
 * @remark They are not defaults or useful for anything else
 */
const dummyOptions: NxArgs = {
  target: '',
  configuration: '',
  runner: '',
  parallel: false,
  maxParallel: 0,
  'max-parallel': 0,
  untracked: false,
  uncommitted: false,
  all: false,
  base: 'base',
  head: 'head',
  exclude: ['exclude'],
  files: [''],
  onlyFailed: false,
  'only-failed': false,
  verbose: false,
  help: false,
  version: false,
  quiet: false,
  plain: false,
  withDeps: false,
  'with-deps': false,
  projects: [],
  select: ''
} as any;

const nxSpecific = Object.keys(dummyOptions);

export interface NxArgs {
  target?: string;
  configuration?: string;
  runner?: string;
  parallel?: boolean;
  maxParallel?: number;
  'max-parallel'?: number;
  untracked?: boolean;
  uncommitted?: boolean;
  all?: boolean;
  base?: string;
  head?: string;
  exclude?: string[];
  files?: string[];
  onlyFailed?: boolean;
  'only-failed'?: boolean;
  verbose?: boolean;
  help?: boolean;
  version?: boolean;
  quiet?: boolean;
  plain?: boolean;
  withDeps?: boolean;
  'with-deps'?: boolean;
  projects?: string[];
  select?: string;
  _: string[];
}

const ignoreArgs = ['$0', '_'];

export function splitArgsIntoNxArgsAndOverrides(
  args: yargs.Arguments
): { nxArgs: NxArgs; overrides: yargs.Arguments } {
  const nxArgs: any = {};
  const overrides = yargsParser(args._);
  delete overrides._;

  Object.entries(args).forEach(([key, value]) => {
    if (nxSpecific.includes(key as any)) {
      nxArgs[key] = value;
    } else if (!ignoreArgs.includes(key)) {
      overrides[key] = value;
    }
  });

  if (!nxArgs.projects) {
    nxArgs.projects = [];
  } else {
    nxArgs.projects = args.projects.split(',').map((p: string) => p.trim());
  }

  return { nxArgs, overrides };
}
