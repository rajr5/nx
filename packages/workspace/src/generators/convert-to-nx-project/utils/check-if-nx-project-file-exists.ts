import { ProjectConfiguration, Tree } from '@nrwl/devkit';
import { getProjectConfigurationPath } from './get-project-configuration-path';

export function checkIfNxProjectFileExists(
  host: Tree,
  project: ProjectConfiguration
) {
  return host.exists(getProjectConfigurationPath(project));
}
