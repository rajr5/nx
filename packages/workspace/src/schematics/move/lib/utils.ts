import { WorkspaceDefinition } from '@angular-devkit/core/src/workspace';
import * as path from 'path';
import { Schema } from '../schema';

/**
 * This helper function ensures that we don't move libs or apps
 * outside of the folders they should be in.
 *
 * This will break if someone isn't using the default libs/apps
 * folders. In that case, they're on their own :/
 *
 * @param schema
 * @param workspace
 */
export function getDestination(
  schema: Schema,
  workspace: WorkspaceDefinition | any
): string {
  const project = workspace.projects.get
    ? workspace.projects.get(schema.projectName)
    : workspace.projects[schema.projectName];
  const projectType = project.extensions
    ? project.extensions['projectType']
    : project.projectType;

  let rootFolder = 'libs';
  if (projectType === 'application') {
    rootFolder = 'apps';
  }
  return path.join(rootFolder, schema.destination).split(path.sep).join('/');
}

/**
 * Replaces slashes with dashes
 *
 * @param path
 */
export function getNewProjectName(path: string): string {
  return path.replace(/\//g, '-');
}

/**
 * Normalizes slashes (removes duplicates)
 *
 * @param input
 */
export function normalizeSlashes(input: string): string {
  return input
    .split('/')
    .filter((x) => !!x)
    .join('/');
}
