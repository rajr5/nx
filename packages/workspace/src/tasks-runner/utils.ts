import { Task } from './tasks-runner';
import { ProjectGraphNode } from '../core/project-graph';

const commonCommands = ['build', 'test', 'lint', 'e2e', 'deploy'];

export function getCommand(cliCommand: string, isYarn: boolean, task: Task) {
  const args = Object.entries(task.overrides || {})
    .map(([prop, value]) => `--${prop}=${value}`)
    .join(' ');

  if (commonCommands.includes(task.target.target)) {
    const config = task.target.configuration
      ? `--configuration ${task.target.configuration} `
      : '';
    return `${cliCommand}${isYarn ? '' : ' --'} ${task.target.target} ${
      task.target.project
    } ${config} ${args}`.trim();
  } else {
    const config = task.target.configuration
      ? `:${task.target.configuration} `
      : '';
    return `${cliCommand}${isYarn ? '' : ' --'}  run ${task.target.project}:${
      task.target.target
    }${config} ${args}`.trim();
  }
}

export function getOutputs(p: Record<string, ProjectGraphNode>, task: Task) {
  const architect = p[task.target.project].data.architect[task.target.target];
  let opts = architect.options || {};
  if (
    architect.configurations &&
    architect.configurations[task.target.configuration]
  ) {
    opts = {
      ...opts,
      ...architect.configurations[task.target.configuration]
    };
  }
  let outputs = [];
  if (opts.outputPath) {
    outputs.push(opts.outputPath);
  }
  return outputs;
}
