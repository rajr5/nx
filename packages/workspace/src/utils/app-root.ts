import { fileExists } from './fileutils';
import * as path from 'path';

export const appRootPath = pathInner(__dirname);

function pathInner(dir: string): string {
  if (path.dirname(dir) === dir) return process.cwd();
  if (
    fileExists(path.join(dir, 'workspace.json')) ||
    fileExists(path.join(dir, 'angular.json'))
  ) {
    // unit test
    if (process.argv[1].indexOf('jest-worker') > -1) {
      return `${dir}/tmp/unit`;
    } else {
      return dir;
    }
  } else {
    return pathInner(path.dirname(dir));
  }
}
