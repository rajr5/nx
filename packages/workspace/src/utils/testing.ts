import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Rule, Tree } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../collection.json')
);

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/workspace/migrations',
  join(__dirname, '../../migrations.json')
);

export function runSchematic(schematicName: string, options: any, tree: Tree) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}

export function runMigration(migrationName: string, options: any, tree: Tree) {
  return migrationTestRunner
    .runSchematicAsync(migrationName, options, tree)
    .toPromise();
}
