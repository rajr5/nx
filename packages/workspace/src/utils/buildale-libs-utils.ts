import {
  ProjectGraph,
  ProjectGraphNode,
  ProjectType
} from '../core/project-graph';
import { BuilderContext } from '@angular-devkit/architect';
import { join } from 'path';
import {
  fileExists,
  readJsonFile,
  writeJsonFile
} from '@nrwl/workspace/src/utils/fileutils';
import { stripIndents } from '@angular-devkit/core/src/utils/literals';

function isBuildable(projectGraph: ProjectGraphNode): boolean {
  return (
    projectGraph.data.architect &&
    projectGraph.data.architect.build &&
    projectGraph.data.architect.build.builder !== ''
  );
}

function getOutputPath(projectGraph: ProjectGraphNode): string {
  const opts = projectGraph.data.architect.build.options;
  if (opts && opts.outputPath) {
    return opts.outputPath;
  } else {
    return `dist/${projectGraph.data.root}`;
  }
}

export type DependentBuildableProjectNode = {
  name: string;
  node: ProjectGraphNode;
};

export function calculateProjectDependencies(
  projGraph: ProjectGraph,
  context: BuilderContext
): { target: ProjectGraphNode; dependencies: DependentBuildableProjectNode[] } {
  const target = projGraph.nodes[context.target.project];
  // gather the library dependencies
  const dependencies = (projGraph.dependencies[context.target.project] || [])
    .map(dependency => {
      const depNode = projGraph.nodes[dependency.target];

      if (depNode.type === ProjectType.lib && isBuildable(depNode)) {
        const libPackageJson = readJsonFile(
          join(context.workspaceRoot, depNode.data.root, 'package.json')
        );

        return {
          name: libPackageJson.name, // i.e. @workspace/mylib
          node: depNode
        };
      } else {
        return null;
      }
    })
    .filter(x => !!x);
  return { target, dependencies };
}

export function checkDependentProjectsHaveBeenBuilt(
  context: BuilderContext,
  projectDependencies: DependentBuildableProjectNode[]
): boolean {
  const depLibsToBuildFirst: DependentBuildableProjectNode[] = [];

  // verify whether all dependent libraries have been built
  projectDependencies.forEach(dep => {
    // check whether dependent library has been built => that's necessary
    const packageJsonPath = join(
      context.workspaceRoot,
      getOutputPath(dep.node),
      'package.json'
    );

    if (!fileExists(packageJsonPath)) {
      depLibsToBuildFirst.push(dep);
    }
  });

  if (depLibsToBuildFirst.length > 0) {
    context.logger.error(stripIndents`
      Some of the project ${
        context.target.project
      }'s dependencies have not been built yet. Please build these libraries before:
      ${depLibsToBuildFirst.map(x => ` - ${x.node.name}`).join('\n')}

      Try: nx run-many --target build --projects ${context.target.project},...
    `);

    return false;
  } else {
    return true;
  }
}

export function updatePaths(
  dependencies: DependentBuildableProjectNode[],
  paths: { [k: string]: string[] }
) {
  dependencies.forEach(depp => {
    if (isBuildable(depp.node)) {
      paths[depp.name] = [getOutputPath(depp.node)];
    }
  });
}

/**
 * Updates the peerDependencies section in the `dist/lib/xyz/package.json` with
 * the proper dependency and version
 */
export function updateBuildableProjectPackageJsonDependencies(
  context: BuilderContext,
  target: ProjectGraphNode,
  dependencies: DependentBuildableProjectNode[]
) {
  // if we have dependencies, update the `dependencies` section of the package.json
  const packageJsonPath = `${getOutputPath(target)}/package.json`;
  const packageJson = readJsonFile(packageJsonPath);

  packageJson.dependencies = packageJson.dependencies || {};

  let updatePackageJson = false;
  dependencies.forEach(entry => {
    if (
      !hasDependency(packageJson, 'dependencies', entry.name) &&
      !hasDependency(packageJson, 'devDependencies', entry.name) &&
      !hasDependency(packageJson, 'peerDependencies', entry.name)
    ) {
      // read the lib version (should we read the one from the dist?)
      const depPackageJsonPath = join(
        context.workspaceRoot,
        getOutputPath(entry.node),
        'package.json'
      );
      const depPackageJson = readJsonFile(depPackageJsonPath);

      packageJson.dependencies[entry.name] = depPackageJson.version;
      updatePackageJson = true;
    }
  });

  if (updatePackageJson) {
    writeJsonFile(packageJsonPath, packageJson);
  }
}

// verify whether the package.json already specifies the dep
function hasDependency(outputJson, depConfigName: string, packageName: string) {
  if (outputJson[depConfigName]) {
    return outputJson[depConfigName][packageName];
  } else {
    return false;
  }
}
