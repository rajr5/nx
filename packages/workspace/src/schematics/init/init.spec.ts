import { Tree } from '@angular-devkit/schematics';
import { runSchematic } from '../../utils/testing';
import { UnitTestTree } from '@angular-devkit/schematics/testing';

describe('workspace', () => {
  let appTree: UnitTestTree;

  beforeEach(() => {
    appTree = new UnitTestTree(Tree.empty());
  });

  it('should error if no package.json is present', async () => {
    try {
      await runSchematic('ng-add', { name: 'myApp' }, appTree);
      fail('should throw');
    } catch (e) {
      expect(e.message).toContain('Cannot find package.json');
    }
  });

  it('should error if no e2e/protractor.conf.js is present', async () => {
    appTree.create('/package.json', JSON.stringify({}));
    appTree.create(
      '/angular.json',
      JSON.stringify({
        projects: {
          proj1: {
            architect: {
              e2e: {
                options: {
                  protractorConfig: 'e2e/protractor.conf.js'
                }
              }
            }
          }
        }
      })
    );

    try {
      await runSchematic('ng-add', { name: 'proj1' }, appTree);
    } catch (e) {
      expect(e.message).toContain(
        'An e2e project was specified but e2e/protractor.conf.js could not be found.'
      );
    }
  });

  it('should error if no angular.json is present', async () => {
    try {
      appTree.create('/package.json', JSON.stringify({}));
      appTree.create('/e2e/protractor.conf.js', '');
      await runSchematic('ng-add', { name: 'myApp' }, appTree);
    } catch (e) {
      expect(e.message).toContain('Cannot find angular.json');
    }
  });

  it('should error if the angular.json specifies more than one app', async () => {
    appTree.create('/package.json', JSON.stringify({}));
    appTree.create('/e2e/protractor.conf.js', '');
    appTree.create(
      '/angular.json',
      JSON.stringify({
        projects: {
          proj1: {},
          'proj1-e2e': {},
          proj2: {},
          'proj2-e2e': {}
        }
      })
    );
    try {
      await runSchematic('ng-add', { name: 'myApp' }, appTree);
    } catch (e) {
      expect(e.message).toContain('Can only convert projects with one app');
    }
  });

  it('should work without nested tsconfig files', async () => {
    appTree.create('/package.json', JSON.stringify({}));
    appTree.create(
      '/angular.json',
      JSON.stringify({
        version: 1,
        defaultProject: 'myApp',
        projects: {
          myApp: {
            root: '',
            sourceRoot: 'src',
            architect: {
              build: {
                options: {
                  tsConfig: 'tsconfig.app.json'
                },
                configurations: {}
              },
              test: {
                options: {
                  tsConfig: 'tsconfig.spec.json'
                }
              },
              lint: {
                options: {
                  tsConfig: 'tsconfig.app.json'
                }
              },
              e2e: {
                options: {
                  protractorConfig: 'e2e/protractor.conf.js'
                }
              }
            }
          }
        }
      })
    );
    appTree.create(
      '/tsconfig.app.json',
      '{"extends": "../tsconfig.json", "compilerOptions": {}}'
    );
    appTree.create(
      '/tsconfig.spec.json',
      '{"extends": "../tsconfig.json", "compilerOptions": {}}'
    );
    appTree.create('/tsconfig.json', '{"compilerOptions": {}}');
    appTree.create('/tslint.json', '{"rules": {}}');
    appTree.create('/e2e/protractor.conf.js', '// content');
    appTree.create('/src/app/app.module.ts', '// content');
    const tree = await runSchematic('ng-add', { name: 'myApp' }, appTree);
    expect(tree.exists('/apps/myApp/tsconfig.app.json')).toBe(true);
  });

  it('should work with nested (sub-dir) tsconfig files', async () => {
    appTree.create('/package.json', JSON.stringify({}));
    appTree.create(
      '/angular.json',
      JSON.stringify({
        version: 1,
        defaultProject: 'myApp',
        projects: {
          myApp: {
            sourceRoot: 'src',
            architect: {
              build: {
                options: {
                  tsConfig: 'src/tsconfig.app.json'
                },
                configurations: {}
              },
              test: {
                options: {
                  tsConfig: 'src/tsconfig.spec.json'
                }
              },
              lint: {
                options: {
                  tsConfig: 'src/tsconfig.app.json'
                }
              },
              e2e: {
                options: {
                  protractorConfig: 'e2e/protractor.conf.js'
                }
              }
            }
          }
        }
      })
    );
    appTree.create(
      '/src/tsconfig.app.json',
      '{"extends": "../tsconfig.json", "compilerOptions": {}}'
    );
    appTree.create(
      '/src/tsconfig.spec.json',
      '{"extends": "../tsconfig.json", "compilerOptions": {}}'
    );
    appTree.create('/tsconfig.json', '{"compilerOptions": {}}');
    appTree.create('/tslint.json', '{"rules": {}}');
    appTree.create('/e2e/protractor.conf.js', '// content');
    appTree.create('/src/app/app.module.ts', '// content');
    const tree = await runSchematic('ng-add', { name: 'myApp' }, appTree);
    expect(tree.exists('/apps/myApp/tsconfig.app.json')).toBe(true);
  });
});
