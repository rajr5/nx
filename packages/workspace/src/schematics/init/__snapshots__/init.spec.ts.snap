// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workspace move to nx layout should update tsconfig.base.json if present 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": ".",
    "paths": Object {},
    "rootDir": ".",
  },
}
`;

exports[`workspace move to nx layout should update tsconfig.json if tsconfig.base.json if present 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": ".",
    "paths": Object {},
    "rootDir": ".",
  },
}
`;
